monai/transforms/croppad/array.py

```diff
@@ -84,63 +84,110 @@ class Pad(Transform):
+    ) -> torch.Tensor:
         """
         Args:
-            img: data to be transformed, assuming `img` is channel-first and
-                padding doesn't apply to the channel dim.
-        mode: available modes for numpy array:{``"constant"``, ``"edge"``, ``"linear_ramp"``, ``"maximum"``,
-            ``"mean"``, ``"median"``, ``"minimum"``, ``"reflect"``, ``"symmetric"``, ``"wrap"``, ``"empty"``}
-            available modes for PyTorch Tensor: {``"constant"``, ``"reflect"``, ``"replicate"`` or ``"circular"``}.
-            One of the listed string values or a user supplied function. Defaults to `self.mode`.
-            See also: https://numpy.org/doc/1.18/reference/generated/numpy.pad.html
-            https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html
+            img: data to be transformed, assuming `img` is channel-first and padding doesn't apply to the channel dim.
+            to_pad: the amount to be padded in each dimension [(low_H, high_H), (low_W, high_W), ...].
+                default to `self.to_pad`.
+            mode: available modes for numpy array:{``"constant"``, ``"edge"``, ``"linear_ramp"``, ``"maximum"``,
+                ``"mean"``, ``"median"``, ``"minimum"``, ``"reflect"``, ``"symmetric"``, ``"wrap"``, ``"empty"``}
+                available modes for PyTorch Tensor: {``"constant"``, ``"reflect"``, ``"replicate"``, ``"circular"``}.
+                One of the listed string values or a user supplied function. Defaults to ``"constant"``.
+                See also: https://numpy.org/doc/1.18/reference/generated/numpy.pad.html
+                https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html
+            kwargs: other arguments for the `np.pad` or `torch.pad` function.
+                note that `np.pad` treats channel dimension as the first dimension.
 
         """
-        if not np.asarray(self.to_pad).any():
-            # all zeros, skip padding
-            return img
-        mode = convert_pad_mode(dst=img, mode=mode or self.mode).value
-        pad = self._pt_pad if isinstance(img, torch.Tensor) else self._np_pad
-        return pad(img, self.to_pad, mode, **self.kwargs)  # type: ignore
+        to_pad_ = self.to_pad if to_pad is None else to_pad
+        if to_pad_ is None:
+            to_pad_ = self.compute_pad_width(img.shape[1:])
+        mode_ = self.mode if mode is None else mode
+        kwargs_ = dict(self.kwargs)
+        kwargs_.update(kwargs)
+
+        img_t = convert_to_tensor(data=img, track_meta=get_track_meta())
+
+        # all zeros, skip padding
+        if np.asarray(to_pad_).any():
+            try:
+                mode_ = convert_pad_mode(dst=img_t, mode=mode_).value
+                out = self._pt_pad(img_t, pad_width=to_pad_, mode=mode_, **kwargs_)
+            # but if mode or args don't exist in pytorch, use numpy instead
+            except (ValueError, TypeError) as err:
```

Comment:
the exception handling will be slow, if we know the user-specified mode is numpy only, perhaps we switch to numpy before the exception is raised? this will be efficient in typical use cases.

Reply:
Hi @wyli ,

That's the same logic from #4536 , I think it's not easy to detect whether the args are for numpy.
For example, both numpy and torch have the `constant` mode, but the `kwargs` can be different.

Thanks.

Reply:
My understanding is for the numpy only modes we can decide before raising an exception.

Reply:
OK, updated in the latest commit.

Thanks.